sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,600 @@
+digital assets within the Stacks blockchain (e.g., BNS names, Stacks
+tokens, etc.)
+
+While application-chains may use any smart-contract language that they
+like, this smart contracting language's VM will be a part of
+blockstack-core, and, as such, any blockstack-core node will be able to
+validate application chains using this smart contracting language with
+a simple configuration change.
+
+This smart contracting language permits static analysis of any legal
+smart contract to determine runtime costs. This smart contracting
+language is not only Turing-incomplete (a requirement for such static
+analysis to be guaranteed successful), but readily permits other kinds
+of proofs to be made about the code as well.
+
+# Design
+
+A smart contract is composed of two parts:
+
+1. A data-space, which is a set of tables of data which only the
+   smart contract may modify
+2. A set of functions which operate within the data-space of the
+   smart contract, though they may call public functions from other smart
+   contracts.
+
+Users call smart contracts' public functions by broadcasting a
+transaction on the blockchain which invokes the public function.
+
+This smart contracting language differs from most other smart
+contracting languages in two important ways:
+
+1. The language _is not_ intended to be compiled. The LISP language
+   described in this document is the specification for correctness.
+2. The language _is not_ Turing complete. This allows us to guarantee
+   that static analysis of programs to determine properties like
+   runtime cost and data usage can complete successfully.
+
+## Specifying Contracts
+
+A smart contract definition is specified in a LISP language with the
+following limitations:
+
+1. Recursion is illegal and there is no `lambda` function.
+2. Looping may only be performed via `map`, `filter`, or `fold`
+3. The only atomic types are booleans, integers, fixed length
+   buffers, and principals
+4. There is additional support for lists of the atomic types, however
+   the only variable length lists in the language appear as function
+   inputs (i.e., there is no support for list operations like append
+   or join).
+5. Variables may only be created via `let` binding and there
+   is no support for mutating functions like `set`.
+6. Defining of constants and functions are allowed for simplifying
+   code using `define` statement. However, these are purely
+   syntactic. If a definition cannot be inlined, the contract will be
+   rejected as illegal. These definitions are also _private_, in that
+   functions defined this way may only be called by other functions
+   defined in the given smart contract.
+7. Functions specified via `define-public` statements are _public_
+   functions. Arguments to these functions must specify their types.
+
+Public functions return a boolean result. If the function returns
+`true`, then it is considered valid, and any changes made to the
```

Comment:
What's `it` referring to here?  The transaction?

Also, to be clear, of the transaction calls `A.foo` and `A.foo` calls `B.bar` (where `A` and `B` are contracts, and `foo` and `bar` are public functions), and `B.bar` returns `true`, but on completion `A.foo` returns `false`, then the effects of `B.bar` are *not* materialized, right?  i.e. changes can only be materialized if the public function invoked by the transaction evaluates to `true`.

Reply:
`it` is referring to the function call.

And yes, that's the case. Updated the text to clarify.